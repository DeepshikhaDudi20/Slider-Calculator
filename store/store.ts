import { configureStore } from '@reduxjs/toolkit'
import loanCalculatorSlice from '../slices/loanCalculatorSlice'

export function makeStore() {
    return configureStore({
        reducer:
            { loanCalculator: loanCalculatorSlice },
    })
}

export const store = makeStore()

export type AppState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch