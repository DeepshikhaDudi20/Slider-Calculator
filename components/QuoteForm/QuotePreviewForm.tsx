import React, { FC, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { FaExchangeAlt } from 'react-icons/fa';
import { FormContainer, ToggleIcon, FieldWithSlider, FieldSliderValue, EditableField, Label } from '../layout/Form';
import { loanCalculator } from '../../slices/loanCalculatorSlice';
import { formData } from '../../AppData/constants';
import { termToYear } from '../../utils/loanCalculator';
import { termOptions } from '../../AppData/constants';
import { AppState } from '../../store/store';
import QuoteFooter from './QuoteFooter';
import Slider from '../Slider';
import Select from '../Select';
import Input from '../Input';

const QuotePreviewForm = () => {
  const [showInput, setShowInput] = useState(false);
  const dispatch = useDispatch();

  const { amount, term, interestRate, monthlyRepayment } = useSelector((state: AppState) => state.loanCalculator);

  const amountFocusHandler = (e: React.FormEvent<HTMLInputElement>) => {
    try{
    let amount = parseInt(e.currentTarget.value);
    if (amount < formData.amountMinValue) amount = formData.amountMinValue;
    if (amount > formData.amountMaxValue) amount = formData.amountMaxValue;
    dispatch(loanCalculator({ amount: amount, term: term }));
    }
    catch (error) {
      console.log(`Error occurred while amount input is focused out ${error}`);
    }
  };

  const amountChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    try{
    dispatch(loanCalculator({ amount: parseInt(e.target.value), term: term }));
    }
    catch (error) {
      console.log(`Error occurred while amount value is changed ${error}`);
    }
  };

  const termChangeHandler = (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLSelectElement>) => {
    try{
    dispatch(loanCalculator({ amount: amount, term: parseInt(e.target.value) }));
    }
    catch (error) {
      console.log(`Error occurred while term value is changed ${error}`);
    }
  };

  return (
    <FormContainer data-testid='quote-form-preview'>
      <ToggleIcon>
        <FaExchangeAlt data-testid='toggle-icon' onClick={() => setShowInput(value => !value)} />
      </ToggleIcon>

      <FieldWithSlider data-testid='amount-field'>
        <Label aria-label="loan-amount-label" htmlFor="loan-amount-slider">
          I want to borrow
          {!showInput && (
            <FieldSliderValue id="loan-amount-value" data-testid='loan-amount'>
              Â£{amount}
            </FieldSliderValue>
          )}
        </Label>
        {!showInput && (
          <Slider
            id="loan-amount-slider"
            min={formData.amountMinValue}
            max={formData.amountMaxValue}
            step={formData.amountStep}
            defaultValue={amount}
            onChangeHandler={amountChangeHandler}
          />
        )}

        {showInput && (
          <EditableField>
            <Input type="number" dataTestId="loan-amount-input" value={amount}
            onChangeHandler={amountChangeHandler}
            onFocusHandler={amountFocusHandler} min={1000} max={20000}            />
          </EditableField>
        )}
      </FieldWithSlider>

      <FieldWithSlider data-testid='term-field'>
        <Label aria-label="loan-term-label" htmlFor="loan-term-slider">
          Over
          {!showInput && (
            <FieldSliderValue id="loan-term-value" data-testid='loan-term'>
              {termToYear(term)}
            </FieldSliderValue>
          )}
        </Label>
        {!showInput && (
          <Slider
            id="loan-term-slider"
            min={formData.termMinValue}
            max={formData.termMaxValue}
            step={formData.termStep}
            defaultValue={term}
            onChangeHandler={termChangeHandler}
          />
        )}

        {showInput &&
          <EditableField>
            <Select name="loan-term" dataTestId="loan-term-input" value={term} options={termOptions}
              onChangeHandler={termChangeHandler} />
          </EditableField>}
      </FieldWithSlider>
      <QuoteFooter interest={interestRate} repayment={monthlyRepayment} />
    </FormContainer>
  );
}

export default React.memo(QuotePreviewForm);
