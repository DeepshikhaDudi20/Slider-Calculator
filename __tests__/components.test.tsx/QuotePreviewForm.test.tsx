/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect'
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../store/store'
import { axe, toHaveNoViolations } from 'jest-axe';
import { create, act } from "react-test-renderer";
import QuotePreviewForm from '../../components/QuoteForm/QuotePreviewForm';

expect.extend(toHaveNoViolations);

describe('The Quote Preview Form Component', () => {
  const component = <Provider store={store}>
    <QuotePreviewForm />
  </Provider>

  test('should render snapshots', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });


  test('should render loan amount and term based on initial state of store', () => {
    render(component);
    expect(screen.getByTestId('loan-amount').textContent).toEqual('£1000')
    expect(screen.getByTestId('loan-term').textContent).toEqual('1 Year')
  })

  test("should render updated loan amount when amount slider value is changed", () => {
    const component = create(<Provider store={store}><QuotePreviewForm /></Provider>);
    const instance = component.root;
    const amountSlider = instance.findByProps({ id: 'loan-amount-slider' });

    act(() => {
      amountSlider.props.onChangeHandler({ target: { value: 2000 } });
    });

    expect(amountSlider.props.defaultValue).toEqual(2000);

    const amountValueElement = instance.findByProps({ id: 'loan-amount-value' });
    // /Get your quote/i
    expect(amountValueElement.props.children[1]).toEqual(2000);
  });

  test("should render updated term when term slider value is changed", async () => {
    const instance = create(component).root;
    const termSlider = instance.findByProps({ id: 'loan-term-slider' });

    await act(() => {
      termSlider.props.onChangeHandler({ target: { value: 24 } });
    });

    expect(termSlider.props.defaultValue).toEqual(24);

    const termValueElement = instance.findByProps({ id: 'loan-term-value' });
    expect(termValueElement.props.children).toEqual('2 Years');
  });

  test("should render updated fractional term when term slider value is changed", () => {
    const instance = create(component).root;
    const termSlider = instance.findByProps({ id: 'loan-term-slider' });

    act(() => {
      termSlider.props.onChangeHandler({ target: { value: 18 } });
    });

    expect(termSlider.props.defaultValue).toEqual(18);

    const termValueElement = instance.findByProps({ id: 'loan-term-value' });
    expect(termValueElement.props.children).toEqual('1½ Years');
  });

  test("should render input elements on clicking toggle icon", async () => {
    render(component);
    expect(screen.queryByTestId('loan-amount-input')).not.toBeInTheDocument()
    expect(screen.queryByTestId('loan-term-input')).not.toBeInTheDocument()

    const toggleIconElement = screen.getByTestId('toggle-icon');
    fireEvent.click(toggleIconElement)
    expect(screen.getByTestId('loan-amount-input')).toBeInTheDocument()
    expect(screen.getByTestId('loan-term-input')).toBeInTheDocument()
  });

  test("should not have any accessibility violations", async () => {
    const { container } = render(component);
    expect(await axe(container)).toHaveNoViolations();
  });
});

