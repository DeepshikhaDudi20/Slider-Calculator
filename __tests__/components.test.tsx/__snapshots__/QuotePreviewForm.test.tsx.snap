// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Quote Preview Form Component should render snapshots 1`] = `
<form
  className="Form__FormContainer-sc-1wozjz1-0 ha-dhje"
  data-testid="quote-form-preview"
>
  <div
    className="Form__ToggleIcon-sc-1wozjz1-1 cdQdtt"
  >
    <svg
      data-testid="toggle-icon"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 168v-16c0-13.255 10.745-24 24-24h360V80c0-21.367 25.899-32.042 40.971-16.971l80 80c9.372 9.373 9.372 24.569 0 33.941l-80 80C409.956 271.982 384 261.456 384 240v-48H24c-13.255 0-24-10.745-24-24zm488 152H128v-48c0-21.314-25.862-32.08-40.971-16.971l-80 80c-9.372 9.373-9.372 24.569 0 33.941l80 80C102.057 463.997 128 453.437 128 432v-48h360c13.255 0 24-10.745 24-24v-16c0-13.255-10.745-24-24-24z"
      />
    </svg>
  </div>
  <div
    className="Form__FieldWithSlider-sc-1wozjz1-2 ipwwtW"
    data-testid="amount-field"
  >
    <label
      aria-label="loan-amount-label"
      className="Form__Label-sc-1wozjz1-3 hwjtqU"
      htmlFor="loan-amount-slider"
    >
      I want to borrow
      <span
        className="Form__FieldSliderValue-sc-1wozjz1-4 fiMSMo"
        data-testid="loan-amount"
        id="loan-amount-value"
      >
        £
        1000
      </span>
    </label>
    <input
      className="Form__SliderInput-sc-1wozjz1-8 eBKLWQ"
      defaultValue={1000}
      id="loan-amount-slider"
      max={20000}
      min={1000}
      onChange={[Function]}
      role="slider"
      step={100}
      type="range"
    />
  </div>
  <div
    className="Form__FieldWithSlider-sc-1wozjz1-2 ipwwtW"
    data-testid="term-field"
  >
    <label
      aria-label="loan-term-label"
      className="Form__Label-sc-1wozjz1-3 hwjtqU"
      htmlFor="loan-term-slider"
    >
      Over
      <span
        className="Form__FieldSliderValue-sc-1wozjz1-4 fiMSMo"
        data-testid="loan-term"
        id="loan-term-value"
      >
        1 Year
      </span>
    </label>
    <input
      className="Form__SliderInput-sc-1wozjz1-8 eBKLWQ"
      defaultValue={12}
      id="loan-term-slider"
      max={60}
      min={12}
      onChange={[Function]}
      role="slider"
      step={6}
      type="range"
    />
  </div>
  <div
    className="Footer__QuoteFormStats-sc-j8bigy-0 bftBsi"
  >
    <div
      className="Footer__QuoteFormStatFields-sc-j8bigy-1 dtHkZx"
    >
      <span
        className="Footer__Value-sc-j8bigy-2 cMOfLq"
        data-testid="interest-rate"
      >
        5
        %
      </span>
      <span
        className="Footer__Label-sc-j8bigy-3 fmjiBb"
      >
        Interest rate
      </span>
    </div>
    <div
      className="Footer__QuoteFormStatFields-sc-j8bigy-1 dtHkZx"
    >
      <span
        className="Footer__Value-sc-j8bigy-2 cMOfLq"
        data-testid="repayment-amount"
      >
        £
        86
      </span>
      <span
        className="Footer__Label-sc-j8bigy-3 fmjiBb"
      >
        Monthly repayment
      </span>
    </div>
  </div>
  <div
    className="Footer__ButtonContainer-sc-j8bigy-4 ilBOR"
  >
    <div
      className="Footer__ButtonWrapper-sc-j8bigy-5 bRPHLH"
    >
      <button
        aria-label="Get your quote"
        className="Footer__Button-sc-j8bigy-6 hhoPwJ"
        data-testid="quote-btn"
        onClick={[Function]}
      >
        Get your quote
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angles-right "
          data-icon="angles-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M470.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 256 265.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160zm-352 160l160-160c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L210.7 256 73.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;
